send only one notification in RemoveAll method of AvaloniaList

vmelnikov opened the pull request:
## What does the pull request do?

This PR optimizes RemoveAll method of AvaloniaList to send only one notification

rstm-sf committed:
Use ControlCatalog.NetCore as an artifact

instead of ControlCatalog.Desktop

MarchingCube committed:
Allow for controlling behavior of CaptionButtons.

MarchingCube committed:
Fix access modifer.

Commit:
change default extend chrome hint

Commit:
Merge remote-tracking branch 'origin/master' into change-default-system-chrome-hints

# Conflicts:
#	src/Avalonia.Controls/ApiCompatBaseline.txt

grokys committed:
Merge branch 'master' into infra/artifacts_controlcatalog

grokys committed:
Added DevToolsOptions.

With options to:

- Set the key gesture
- Show as child window
- Set the initial size

Co-Authored-By: workgroupengineering <workgroupengineering@users.noreply.github.com>

kekekeks committed:
Allow binding to classes with Classes.ClassName

kekekeks committed:
Simplify the code a bit

maxkatz6 committed:
Fix docs typo

vmelnikov committed:
send only one notification in RemoveAll method of AvaloniaList

vmelnikov committed:
send only one notification in RemoveAll method of AvaloniaList

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=5728) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=5728) before we can accept your contribution.<br/><br/>:x: vmelnikov [sign now](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=5728)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/AvaloniaUI/Avalonia?pullRequest=5728) it.</sub>

vmelnikov committed:
Merge branch 'master' into remove_all

jmacato committed:
Merge branch 'master' into binding-to-classes

grokys committed:
Fix pointer interaction with reversed direction slider.

Previously dragging a slider with `IsDirectionRevered = true` resulted in the slider moving the wrong way.

jmacato committed:
Merge pull request #5729 from AvaloniaUI/fixes/slider-reversed-mouse

Fix pointer interaction with reversed direction slider.

jmacato committed:
Add Deterministic XamlX ID Generator (#5684)

* Add Deterministic XamlX ID Generator

* Apply suggestions from code review

* simplify stuff and apply review

* Update src/Avalonia.Build.Tasks/XamlCompilerTaskExecutor.cs

* Update src/Avalonia.Build.Tasks/XamlCompilerTaskExecutor.cs

* add the det id gen to runtine xamlx compiler

* rerun tests

* rerun tests

* try this

* use id gen instead of guid

* a

* Update src/Markup/Avalonia.Markup.Xaml.Loader/AvaloniaXamlIlRuntimeCompiler.cs

* Update AvaloniaXamlIlCompilerConfiguration.cs

revert

* Update XamlIlClrPropertyInfoHelper.cs

* Update AvaloniaXamlIlRuntimeCompiler.cs

* fix

* revert hack

* make id gen optional

joaompneves committed:
Show and hide native control if visibility of any of its ancestors changes

Commit:
Merge branch 'master' into change-default-system-chrome-hints

Commit:
Merge pull request #5732 from joaompneves/nativehostcontrol-root-visiblity-fix

Update native control visibility when ancestor visibility changes

Commit:
Merge branch 'master' into change-default-system-chrome-hints

Commit:
maintain the client size when switching between extended and non extended client areas.

Previously the client area would grow to fill the window.

Commit:
whitespace.

Commit:
whitespace.

Commit:
Merge pull request #5642 from AvaloniaUI/change-default-system-chrome-hints

Change default system chrome hints

Commit:
Merge branch 'master' into extended-client-area-win32-maintain-size

maxkatz6 committed:
Merge branch 'master' into feature/devtoolsoptions

maxkatz6 committed:
Merge branch 'master' into captionbuttons-api

maxkatz6 committed:
Merge branch 'master' into infra/artifacts_controlcatalog

maxkatz6 committed:
Merge pull request #5669 from AvaloniaUI/feature/devtoolsoptions

Added DevToolsOptions.

maxkatz6 committed:
Merge branch 'master' into captionbuttons-api

maxkatz6 committed:
Merge pull request #5555 from MarchingCube/captionbuttons-api

Allow for controlling behavior of CaptionButtons.

maxkatz6 committed:
Merge branch 'master' into infra/artifacts_controlcatalog

maxkatz6 committed:
Merge pull request #5481 from rstm-sf/infra/artifacts_controlcatalog

Use ControlCatalog.NetCore as an artifact

jmacato committed:
Merge branch 'master' into binding-to-classes

jmacato committed:
Merge pull request #5710 from AvaloniaUI/binding-to-classes

[WIP] Added `<Button Classes.myClass="{Binding ...}" />` syntax

Commit:
Merge remote-tracking branch 'origin/master' into extended-client-area-win32-maintain-size

Commit:
take into account render scaling.

Commit:
take into account the extended mode when Resize is called.

Commit:
get rid of dodgy logo... sorry

Commit:
Merge pull request #5737 from AvaloniaUI/danwalmsley-patch-1

get rid of dodgy logo... sorry

MarchingCube commented on src/Avalonia.Base/Collections/AvaloniaList.cs:
```diff
@@ -480,12 +480,16 @@ public virtual bool Remove(T item)
             Contract.Requires<ArgumentNullException>(items != null);
-
+            
+            var list = new List<T>();
```
It would be beneficial to create a benchmark for this. I wonder if we should presize `list` if we know that `items` is an `IList` with a known size.

Gillibald replied:
Why do we need `IEnumerable<T>` here can't we use `IList<T>` instead. No need for type checks.

vmelnikov replied:
>  I wonder if we should presize list if we know that items is an IList with a known size.

 I can't garantee that all deleting items are exists in `_inner`. So I can't presize `list` anyway

vmelnikov replied:
> Why do we need `IEnumerable<T>` here can't we use `IList<T>` instead. No need for type checks.

I think this may be useful in some cases. For example we can use **linq** result as `items` parameter without calling `ToList()` method

kekekeks commented on src/Avalonia.Base/Collections/AvaloniaList.cs:
```diff
@@ -480,12 +480,16 @@ public virtual bool Remove(T item)
             }
+            
+            if (list.Count > 0)
+                NotifyRemove(list, -1);
```
-1 is not a valid value for NewStartingIndex / OldStartingIndex. You need to have one notification per sequential range of removed items.

Commit:
Merge branch 'master' into extended-client-area-win32-maintain-size

Commit:
Merge pull request #5736 from AvaloniaUI/extended-client-area-win32-maintain-size

maintain the client size when switching between extended and non exte…

vmelnikov committed:
send one notification per sequential range of removed items in RemoveAll method

vmelnikov committed:
Merge remote-tracking branch 'origin/remove_all' into remove_all

# Conflicts:
#	src/Avalonia.Base/Collections/AvaloniaList.cs
